README.md
# Vertex2NodeByAtt
QGIS plugin : Move lines on points by attribute (and vice versa)

Description
===========
**Vertex2NodeByAtt** is a QGIS plugin that modifies the position of the vertices (the start-point and the end-point) of lines to put them at the position of 2 node objects, thanks to 1 attribute of the line that contains the 2 ids of the 2 nodes objects, with a separator. The reverse is also possible: move points on lines.

French description
==================
**Vertex2NodeByAtt** est une extension QGIS permettant de modifier la position des sommets de lignes (premier point et dernier point) pour les repositionner à l'emplacement de 2 objets ponctuels, grâce à la valeur d'un attribut correspondant à l'identifiant de chacun de ces 2 points (avec un séparateur spécifique entre les 2 identifiants dans la valeur d'attribut). L'inverse est également possible: déplacer les points sur les lignes.

Prerequisite
============
* QGIS 3.10 LTR

Documentation
=============
See the video (in French): https://youtu.be/kHRJkvIrJzo

Example of use:
- the layer LINES has an attribute named REGIDS
- the layer NODES has an attribute named PTID
- a LINES object has the value 'R1-R2' in REGIDS
- a NODES object has the value 'R1' in PTID
- another NODES object has the value 'R2' in PTID
- then the result is:
  1. the start-point of the LINES object is moved to the first NODES object position (R1)
  2. the end-point of the LINES object is moved to the second NODES object position(R2)
  3. the reverse is possible (the NODES objects are moved to the position of the start and end points of the LINES object)

Author
======
* Etienne MORO
* e-mail: em@sigmoe.fr
* website: www.sigmoe.fr

License
=======
GNU Public License (GPL) Version 3



CHANGELOG.md
# CHANGELOG

### 3.3.1 - 2021/09/11

* Avoid exiting when  NULL values

### 3.3.0 - 2020/02/17

* First version

__init__.py
# -*- coding: utf-8 -*-

"""
    ***************************************************************************
    * Plugin name:   SgmVertex2NodeByAtt
    * Plugin type:   QGIS 3 plugin
    * Module:        Initialization
    * Description:   Modify the placement of the vertices (the start-point and
    *                the end-point) of lines to put them at the placement  
    *                of 2 node objects, thanks to 1 attribute of the line
    *                that contains the 2 ids of the 2 nodes objects, with
    *                a separator.
    *                For example:
    *                   - the layer LINES has an attribute named REGIDS
    *                   - the layer NODES has an attribute named PTID
    *                   - a LINES object has the value 'R1-R2' for REGIDS
    *                   - a NODES object has the value 'R1' for PTID
    *                   - another NODES object has the value 'R2' for PTID
    *                   - then the result is:
    *                       1) the start-point of the LINES object is moved
    *                          to the first NODES object position
    *                       2) the end-point of the LINES object is moved
    *                          to the second NODES object position
    *                       3) the reverse is possible (the NODES objects
    *                          are moved to the position of the start and
    *                          end points of the LINES object)
    * Specific lib:  None
    * First release: 2017-05-11
    * Last release:  2020-04-06
    * Copyright:     (C)2020 SIGMOE
    * Email:         em at sigmoe.fr
    * License:       GPL v3
    ***************************************************************************
    * This program is free software: you can redistribute it and/or modify
    * it under the terms of the GNU General Public License as published by
    * the Free Software Foundation, either version 3 of the License, or
    * (at your option) any later version.
    *
    * This program is distributed in the hope that it will be useful,
    * but WITHOUT ANY WARRANTY; without even the implied warranty of
    * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    * GNU General Public License for more details.
    *
    * You should have received a copy of the GNU General Public License
    * along with this program. If not, see <http://www.gnu.org/licenses/>.
    ***************************************************************************
   
    This script initializes the plugin, making it known to QGIS.
"""

def classFactory(iface):
    # load SgmVertex2NodeByAtt class from file sgm_vertex2nodebyatt
    from .sgm_vertex2nodebyatt import SgmVertex2NodeByAtt
    return SgmVertex2NodeByAtt(iface)


sgm_vertex2nodebyatt_rsc.py
# -*- coding: utf-8 -*-

# Resource object code
#
# Created by: The Resource Compiler for PyQt5 (Qt v5.11.2)
#
# WARNING! All changes made in this file will be lost!

from PyQt5 import QtCore

qt_resource_data = b"\
\x00\x00\x01\xfc\
\x89\
\x50\x4e\x47\x0d\x0a\x1a\x0a\x00\x00\x00\x0d\x49\x48\x44\x52\x00\
\x00\x00\x18\x00\x00\x00\x18\x08\x06\x00\x00\x00\xe0\x77\x3d\xf8\
\x00\x00\x01\xc3\x49\x44\x41\x54\x48\xc7\xb5\x94\x3f\x48\x1b\x61\
\x18\xc6\x7f\x77\x86\xd6\x0c\x2d\x5d\x4e\xc4\x88\x58\x5a\x08\x86\
\x86\x58\xca\x29\xb7\xa8\x04\x41\x29\x9d\x1c\x85\x2e\x5a\x70\x14\
\x27\x87\xce\x4e\x0e\x2e\x5d\x8a\x66\x12\x5c\x2a\xb7\x28\x81\x2c\
\x82\x0e\x0a\x8a\xc4\x12\xa8\x76\x28\xad\xcb\x51\x62\x07\x41\xee\
\x92\x08\xf2\x3a\xe4\x2e\x3d\xc3\x89\x46\xef\x1e\xf8\xf8\xfe\x0c\
\xef\xef\x7b\x9f\xe7\xe3\x03\x57\xa2\x69\x42\x04\x52\x89\x58\xaa\
\x68\x9a\x78\xb7\xf7\xd6\xba\x69\x25\xc3\x02\x28\x00\xba\x69\x19\
\xfb\x33\x99\xdd\x81\xaf\xdf\xbf\x00\x49\xe0\x35\xf0\x12\x38\x04\
\x7e\xfa\xc7\xc1\x44\xd7\xd1\x5d\x45\x45\xd3\x44\x39\x3b\x53\x1a\
\x80\xe6\x43\x17\xfa\x04\x48\x03\x29\x17\xea\x81\xfb\x9b\xa1\xc0\
\x11\xf0\xe3\x60\xa2\xab\x72\x2b\xa0\x15\xe9\xa6\x95\xf2\x01\x93\
\x2e\x34\xb5\x3f\x93\x89\x47\x1a\xa8\x6e\x5a\x09\xdd\xb4\xb2\x6e\
\x8e\x3d\x37\x3a\x58\xd8\x2a\xcb\xaf\x73\x78\xf5\x02\x3e\x67\x3b\
\x94\xc7\x80\xfc\x16\x35\x9e\xe9\x49\xb9\xc6\xe0\xa0\xc6\x49\xb9\
\xf6\xf8\x97\xe3\xcb\x32\xe6\x2d\x2e\x2b\x55\xfe\x5d\xd6\xe7\x87\
\x2a\xc8\x85\x46\x07\x35\xc7\xe1\xb7\x5d\x9f\x1f\xaa\x20\x17\x62\
\x73\xab\x45\x39\xfe\x5b\xa1\xea\xd8\x9c\xda\xd0\xe6\xd8\x8c\x2f\
\xee\x4a\x5f\x67\x9c\xa5\x8f\x6f\x5b\xca\x22\xc8\x05\x35\x93\x68\
\xa7\x6a\xdb\x3c\x1f\x35\x38\x75\x40\x1d\x31\xa8\xda\x36\x99\x44\
\x7b\xcb\x1d\x04\xb9\xa0\x00\xe4\x0a\x25\x59\xd9\xfa\xc3\xc5\xfb\
\x0f\x3c\xcb\x6f\xf2\x29\xdb\xcb\xf4\x58\xfa\xde\xb7\xf7\x5c\x00\
\xb8\x1a\x32\x68\xdb\xd9\x03\xa0\xaf\x33\x5e\xcf\x60\x7a\x2c\xad\
\x4c\x0d\x77\xf3\xd4\x5c\x67\x6a\xb8\xbb\xa5\xe2\x00\x61\xba\x70\
\xab\x72\x85\x92\x18\xf3\x1b\xf2\x66\x5b\xc4\x98\xdf\x90\x5c\xa1\
\x14\xfe\xf7\xbf\x9c\x2f\xca\xbb\xd9\x6f\xb2\x9c\x2f\xca\x8d\x0c\
\x64\x72\xf2\xff\xc1\xda\x9a\x12\x26\x54\xf1\x17\x8f\x02\x12\x0b\
\xdb\xa6\x66\x37\xd4\xa8\x8a\x7b\x7b\xb5\xd9\x8e\xd0\x33\x88\xb2\
\x83\xd0\x01\x41\x19\x5c\x03\x1d\x99\xe3\xeb\x3f\x2f\xfc\x5f\x00\
\x00\x00\x00\x49\x45\x4e\x44\xae\x42\x60\x82\
"

qt_resource_name = b"\
\x00\x05\
\x00\x6f\xa6\x53\
\x00\x69\
\x00\x63\x00\x6f\x00\x6e\x00\x73\
\x00\x0e\
\x05\x27\x21\xb4\
\x00\x73\
\x00\x67\x00\x6d\x00\x5f\x00\x76\x00\x74\x00\x78\x00\x32\x00\x6e\x00\x64\x00\x62\x00\x61\x00\x74\x00\x74\
"

qt_resource_struct_v1 = b"\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x01\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\
\x00\x00\x00\x10\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\
"

qt_resource_struct_v2 = b"\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x01\
\x00\x00\x00\x00\x00\x00\x00\x00\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\
\x00\x00\x00\x00\x00\x00\x00\x00\
\x00\x00\x00\x10\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\
\x00\x00\x01\x5d\xa3\xcf\xc9\x62\
"

qt_version = [int(v) for v in QtCore.qVersion().split('.')]
if qt_version < [5, 8, 0]:
    rcc_version = 1
    qt_resource_struct = qt_resource_struct_v1
else:
    rcc_version = 2
    qt_resource_struct = qt_resource_struct_v2

def qInitResources():
    QtCore.qRegisterResourceData(rcc_version, qt_resource_struct, qt_resource_name, qt_resource_data)

def qCleanupResources():
    QtCore.qUnregisterResourceData(rcc_version, qt_resource_struct, qt_resource_name, qt_resource_data)

qInitResources()


i18n/SgmVertex2NodeByAtt_fr.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="fr" sourcelanguage="en">
<context>
    <name>SgmVertex2NodeByAtt</name>
    <message>
        <location filename="../sgm_vertex2nodebyatt.py" line="85"/>
        <source>SIGMOE</source>
        <translation>Sigmoé</translation>
    </message>
    <message>
        <location filename="../sgm_vertex2nodebyatt.py" line="86"/>
        <source>Moves lines to points by attribute</source>
        <translation>Déplace lignes sur points par attribut</translation>
    </message>
    <message>
        <location filename="../sgm_vertex2nodebyatt.py" line="87"/>
        <source>Process stopped</source>
        <translation>Arrêt du traitement</translation>
    </message>
    <message>
        <location filename="../sgm_vertex2nodebyatt.py" line="88"/>
        <source>Move completed</source>
        <translation>Déplacement terminé</translation>
    </message>
    <message>
        <location filename="../sgm_vertex2nodebyatt.py" line="89"/>
        <source>Layer %s is not editable</source>
        <translation>La couche %s n&apos;est pas modifiable</translation>
    </message>
    <message>
        <location filename="../sgm_vertex2nodebyatt.py" line="90"/>
        <source>Field %s is not usable</source>
        <translation>Le champ %s n&apos;est pas exploitable</translation>
    </message>
</context>
<context>
    <name>dlgParams</name>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="26"/>
        <source>Moving lines vertices to points</source>
        <translation>Déplacement des sommets de lignes sur les points (ou inversement)</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="80"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-style:italic;&quot;&gt;QGIS plugin developped by SIGMOÃ‰&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-style:italic;&quot;&gt;Extension QGIS développée par SIGMOÃ‰&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="111"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;a href=&quot;https://sigmoe.fr&quot;&gt;&lt;span style=&quot; text-decoration: underline; color:#0000ff;&quot;&gt;www.sigmoe.fr&lt;/span&gt;&lt;/a&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;a href=&quot;https://sigmoe.fr&quot;&gt;&lt;span style=&quot; text-decoration: underline; color:#0000ff;&quot;&gt;www.sigmoe.fr&lt;/span&gt;&lt;/a&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="160"/>
        <source>OK</source>
        <translation>OK</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="198"/>
        <source>Line layer containing the lines :</source>
        <translation>Couche linéaire contenant les lignes :</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="226"/>
        <source>Lines field containing the reference of the attached points :</source>
        <translation>Champ des lignes contenant la référence des points rattachés :</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="254"/>
        <source>Reference separator in the lines field :</source>
        <translation>Séparateur des références dans le champ des lignes :</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="289"/>
        <source>Point layer containing the points :</source>
        <translation>Couche ponctuelle contenant les points :</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="317"/>
        <source>Points field containing the reference :</source>
        <translation>Champ des points contenant la référence :</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="345"/>
        <source>Object movement mode :</source>
        <translation>Mode de déplacement des objets :</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="358"/>
        <source>the lines are moved to the points</source>
        <translation>les lignes sont déplacées sur les points</translation>
    </message>
    <message>
        <location filename="../gui/sgm_vertex2nodebyatt_dlgparams.ui" line="374"/>
        <source>the points are moved along the line</source>
        <translation>les points sont déplacés sur les lignes</translation>
    </message>
</context>
</TS>


sgm_vertex2nodebyatt.py
# -*- coding: utf-8 -*-

"""
    ***************************************************************************
    * Plugin name:   SgmVertex2NodeByAtt
    * Plugin type:   QGIS 3 plugin
    * Module:        Main
    * Description:   Modify the placement of the vertices (the start-point and
    *                the end-point) of lines to put them at the placement  
    *                of 2 node objects, thanks to 1 attribute of the line
    *                that contains the 2 ids of the 2 nodes objects, with
    *                a separator.
    *                For example:
    *                   - the layer LINES has an attribute named REGIDS
    *                   - the layer NODES has an attribute named PTID
    *                   - a LINES object has the value 'R1-R2' for REGIDS
    *                   - a NODES object has the value 'R1' for PTID
    *                   - another NODES object has the value 'R2' for PTID
    *                   - then the result is:
    *                       1) the start-point of the LINES object is moved
    *                          to the first NODES object position
    *                       2) the end-point of the LINES object is moved
    *                          to the second NODES object position
    *                       3) the reverse is possible (the NODES objects
    *                          are moved to the position of the start and
    *                          end points of the LINES object)
    * Specific lib:  None
    * First release: 2017-05-11
    * Last release:  2020-04-30
    * Copyright:     (C)2020 SIGMOE
    * Email:         em at sigmoe.fr
    * License:       GPL v3
    ***************************************************************************
    * This program is free software: you can redistribute it and/or modify
    * it under the terms of the GNU General Public License as published by
    * the Free Software Foundation, either version 3 of the License, or
    * (at your option) any later version.
    *
    * This program is distributed in the hope that it will be useful,
    * but WITHOUT ANY WARRANTY; without even the implied warranty of
    * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    * GNU General Public License for more details.
    *
    * You should have received a copy of the GNU General Public License
    * along with this program. If not, see <http://www.gnu.org/licenses/>.
    ***************************************************************************
"""

from qgis.PyQt.QtCore import QCoreApplication, QSettings, QTranslator
from qgis.PyQt.QtWidgets import QAction, QMenu
from qgis.PyQt.QtGui import QIcon
from qgis.core import QgsProject
from qgis.utils import iface

import os

from .sgm_vertex2nodebyatt_tr import *
from . import sgm_vertex2nodebyatt_rsc

class SgmVertex2NodeByAtt:
    
    # Initialisation
    def __init__(self, iface):
        self.iface = iface
        self.canvas = iface.mapCanvas()
        self.project = QgsProject.instance()
        self.plugin_dir = os.path.dirname(__file__)
        # initialize locale
        locale = QSettings().value('locale/userLocale')[0:2]
        locale_path = os.path.join(
            self.plugin_dir,
            'i18n',
            'SgmVertex2NodeByAtt_{}.qm'.format(locale))

        if os.path.exists(locale_path):
            self.translator = QTranslator()
            self.translator.load(locale_path)
            QCoreApplication.installTranslator(self.translator)
            
        self.settings = QSettings()
        self.sgm_menu = None
        self.vtx_nodes_tr = None
        
        # Global texts
        self.mnu_title_txt = self.tr("SIGMOE")
        self.fnc_title_txt = self.tr("Moves lines to points by attribute")
        self.err_title_txt = self.tr("Process stopped")
        self.end_txt = self.tr("Move completed")
        self.err_noedit_txt = self.tr("Layer %s is not editable")
        self.err_noatt_txt = self.tr("Field %s is not usable")
        self.gbl_text = [self.fnc_title_txt, self.err_title_txt, self.end_txt, self.err_noedit_txt, self.err_noatt_txt]
    
    def tr(self, message):
        """Get the translation for a string using Qt translation API.

        We implement this ourselves since we do not inherit QObject.

        :param message: String for translation.
        :type message: str, QString

        :returns: Translated version of message.
        :rtype: QString
        """
        # noinspection PyTypeChecker,PyArgumentList,PyCallByClass
        return QCoreApplication.translate('SgmVertex2NodeByAtt', message)

    # Function to add menus
    def sgm_add_submenu(self, submenu):
        if self.sgm_menu != None:
            self.sgm_menu.addMenu(submenu)
        else:
            self.iface.addPluginToMenu("&Sigmoe", submenu.menuAction())
            
    # Initialisation of the menu and the toolbar
    def initGui(self):
        # Add Sigmoe to QGIS menu
        self.sgm_menu = QMenu(QCoreApplication.translate("SigmoeVtxto2NdByatt", self.mnu_title_txt))
        self.iface.mainWindow().menuBar().insertMenu(self.iface.firstRightStandardMenu().menuAction(), self.sgm_menu)
        
        # Add Sigmoe toolbar
        self.toolbar = self.iface.addToolBar(self.mnu_title_txt)
        self.toolbar.setObjectName("SigmoeVtxto2NdByattTB")
        
        # Create actions
        icon = QIcon(r":/icons/sgm_vtx2ndbatt")
        self.transfo_action = QAction(icon, self.fnc_title_txt, self.iface.mainWindow())
               
        # Add actions to the toolbar
        self.toolbar.addActions([self.transfo_action])
        
        # Add actions to the menu
        self.sgm_menu.addActions([self.transfo_action])
        
        # Manage signals
        self.transfo_action.triggered.connect(self.transfo)
                
    # Unload actions
    def unload(self):
        if self.sgm_menu != None:
            self.iface.mainWindow().menuBar().removeAction(self.sgm_menu.menuAction())
            self.sgm_menu.deleteLater()
            self.iface.mainWindow().removeToolBar(self.toolbar)
        else:
            self.iface.removePluginMenu("&SigmoeVtxto2NdByatt", self.sgm_menu.menuAction())
            self.sgm_menu.deleteLater()
        
    # Main function
    def transfo(self):
        '''
            Launch the transformation with the possibility of setting the parameters
        '''
        self.vtx_nodes_tr = Vertex2NodeByAttTr (
                                self.iface,
                                self.canvas,
                                self.project,
                                self.plugin_dir,
                                self.gbl_text
                                )
        self.vtx_nodes_tr.vtx_nodes_modify()


        
